### DISCUSSION TITLE
Using test runners in VS2019 / Rider IDEs

### DISCUSSION THREAD

**Discussion by FrankRay78** [nunit/nunit#3861] [Q&A] [answered]
Hello,

Is it possible / does anyone use the test runners built into either the VS2019 or JB Rider IDEs?

I've followed the various wiki / md instructions and can build the latest NUnit framework locally from powershell (ie. pull down the nunit master and then run 'build.ps1').

But within an IDE, I cannot for the life of me right click on a unit test, and have it run from within the IDE.  I've tried all sorts of settings and framework tweaks to no avail.

![image](https://user-images.githubusercontent.com/52075808/119894545-9f7e6f00-bf34-11eb-9532-8557e1f0ac13.png)

Rider cannot recognise the tests, neither can I get VS2019 to. I've tried both VS2019 and JB Rider IDEs, on Windows and Rider one Windows and Linux.

So... is this really the case (my guess would an issue with multi-framework targeting proj)?  

If so, then how do you write / run / debug your framework unit tests?

Are you literally running and full build and test on every change?

> build.ps1 --target=Test --configuration=Debug

Given there seems to be no way to tell the build script to just run a single given test, I'm slightly at a loss as to how best to setup my dev environment when making nunit framework changes.

I've scoured the docs and I'm none the wiser.  Could someone illuminate my understanding please?

Regards,
Frank

cc @mikkelbu - do you know about this perhaps?

**Comment by mikkelbu** [comment]
Hi @FrankRay78 

I'm sorry, but I don't think I can help you much, as it has been some time since I last did any major development in this project (most of my free time is spent reviewing and triaging issues rather than programming) - and last time I actually depended on the CI to run the tests as I was having some local computer problems with VS prereleases :). Also, it has been a couple of years since I used Resharper and I don't know much about their test runner.

We have some issues which are related to your question #2973 (Write up recommended way to test and link it from the readme) and #3008 (Get codebase tests working with ReSharper and Ride), but perhaps if you add the NUnitAdapter to the relevant test project then you could run the tests from VS. At least from #2973 it sounds like it is possible to select the relevant tests.

**Reply by FrankRay78** [reply]
Thank you @mikkelbu for your reply. 

I'm going to spend some time reading through the issues you mention and see if they help me / I can help in some regard. 

I've spent quite a bit more time fiddling around with tests on both VS and Rider and still cannot get it working from within either IDE (although it builds and tests run ok across all target frameworks from the command line / terminal).

Perhaps something in those issues may help me progress this. Otherwise I think I plan to [log an issue with JetBrains](https://rider-support.jetbrains.com/hc/en-us/articles/115001124744-Problems-with-NET-unit-testing) and see if they can help sort it out, if for no other reason than for my own curiosity. I will feedback anything that may help.

[PS. my interest in collaborating on NUnit originally came from thinking about [fixing this properly](https://github.com/nunit/nunit/pull/1709). Not being able to run tests in the IDE makes the developer experience harder, that's all]

Frank

**Reply by FrankRay78** [reply]
Hello @stevenaw, @hyp-do, @pakrym, @Crown0815, @mvdfugro, @Temtaime, @skomialek

Could you kindly tell me if you have successfully got unit test execution for the nunit.framework.tests project running, in either VS2019 and/or Rider?

If so, did you have to do anything over and above cloning the repo into a new directory?

Also see this issue I've raised if you have a minute #3867

**Reply by mvdfugro** [reply]
The only method I've gotten to work reliably (after having spent maybe 15 minutes on it ;-)) is to update the run configuration to use nunitlits-runner.exe:

```
diff --git a/src/NUnitFramework/tests/Properties/launchSettings.json b/src/NUnitFramework/tests/Properties/launchSettings.json
index 3e530843b..e1cfb11a3 100644
--- a/src/NUnitFramework/tests/Properties/launchSettings.json
+++ b/src/NUnitFramework/tests/Properties/launchSettings.json
@@ -2,8 +2,8 @@
   "profiles": {
     "nunit.framework.tests": {
       "commandName": "Executable",
-      "executablePath": "C:\\src\\nunit\\nunit\\packages\\NUnit.ConsoleRunner.3.7.0\\tools\\nunit3-console.exe",
-      "commandLineArgs": "--inprocess --debug --wait $(SolutionDir)\\bin\\$(Configuration)\\$(TargetFramework)\\nunit.framework.tests.dll"
+      "executablePath": "$(SolutionDir)\\bin\\$(Configuration)\\$(TargetFramework)\\nunitlite-runner.exe",
+      "commandLineArgs": "$(SolutionDir)\\bin\\$(Configuration)\\$(TargetFramework)\\nunit.framework.tests.dll --test=NUnit.Framework.Attributes.LifeCycleAttributeTests"
     }
   }
-}
\ No newline at end of file
+}
```

The --test flag then allows filtering to relevant tests. This also works in debug mode so you can set breakpoints etc.

**Comment by OsirisTerje** [comment, marked as answer]
For anyone dropping in here, this issue was fixed March 25th 2023, in PR #4315 .
